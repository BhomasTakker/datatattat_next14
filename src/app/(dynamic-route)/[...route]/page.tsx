import { getPage } from "@/actions/page/page-actions";
import { IPage } from "@/types/page";
import { PageDisplay } from "@/components/page/page-display";
import styles from "../../page.module.scss";
import { generateMetaDataFromPage } from "@/lib/metadata/generate-metadata";
import { initialiseServices } from "@/lib/services/intialise-services";

type Params = Promise<{ route: string[] }>;
type Props = {
	params: Params;
};

export const revalidate = 960;

// export async function generateStaticParams() {
// 	return [
// 		{ route: ["world"] },
// 		{ route: ["us"] },
// 		{ route: ["uk"] },
// 		{ route: ["ukraine"] },
// 		{ route: ["trade-war"] },
// 	];
// }

export const generateMetadata = async ({ params }: Props) => {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const metadata = await generateMetaDataFromPage(`/${joined}`);
	return metadata;
};

export default async function Page({ params }: Props) {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const page = (await getPage(`/${joined}`)) as IPage;

	if (!page) {
		// omg do something
		return <h1>404</h1>;
	}

	return (
		<div className={styles.page}>
			<PageDisplay page={page} />
		</div>
	);
}
