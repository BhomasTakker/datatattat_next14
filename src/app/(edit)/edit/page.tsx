import { redirect } from "next/navigation";
import { PATHS } from "@/lib/routing/paths";
import isSignupComplete from "@/actions/signup/signup-completed";
import isValidSession from "@/actions/auth/check-session";
import { isValidUser } from "@/actions/auth/check-valid-user";
import { connectToMongoDB } from "@/lib/mongo/db";

// export const dynamic = "force-dynamic";

export default async function Page() {
	await connectToMongoDB();
	await isValidSession();
	await isSignupComplete();
	const { username } = await isValidUser();

	const userHome = `${PATHS.user(username)}`;

	redirect(`${PATHS.edit()}${userHome}`);
}
