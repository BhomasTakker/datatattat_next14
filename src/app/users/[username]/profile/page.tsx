import styles from "../../../page.module.scss";
import { getUserByUsername } from "@/lib/mongo/actions/user";
import { ProfilePage } from "@/components/profile/profile-page";
import { ProfileEditPage } from "@/components/profile/profile-edit-page";
import { getUser } from "@/actions/user/get-user";
import { redirect } from "next/navigation";
import { PATHS } from "@/lib/routing/paths";
import { MainHeader } from "@/components/header/main-header";
import { connectToMongoDB } from "@/lib/mongo/db";

export default async function UserProfile({
	params,
}: {
	params: Promise<{ username: string }>;
}) {
	const { username } = await params;
	const route = ["users", username];

	await connectToMongoDB();
	const sessionUser = await getUser();
	const profileUser = await getUserByUsername(username);

	if (sessionUser && profileUser) {
		if (sessionUser._id.toString() === profileUser._id.toString()) {
			return (
				<>
					<MainHeader route={route} />
					<div className={styles.page}>
						<ProfileEditPage />
					</div>
				</>
			);
		}
	}

	if (!profileUser) {
		// log and redirect
		redirect(PATHS.home());
	}

	return (
		<>
			<MainHeader route={route} />
			<div className={styles.page}>
				<ProfilePage
					user={{ username: profileUser.username, avatar: profileUser.avatar }}
				/>
			</div>
		</>
	);
}
