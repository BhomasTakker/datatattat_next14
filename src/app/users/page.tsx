import { PATHS } from "@/lib/routing/paths";
import { redirect } from "next/navigation";

export default async function Page() {
	// just redirect for now
	redirect(`${PATHS.home()}`);
}
