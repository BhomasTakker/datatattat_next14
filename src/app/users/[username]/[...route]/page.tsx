import { getPage } from "@/actions/page/page-actions";
import { IPage } from "@/types/page";
import { PageDisplay } from "@/components/page/page-display";
import styles from "../../../page.module.scss";
import { PATHS } from "@/lib/routing/paths";

export default async function UserHome({
	params,
}: {
	params: Promise<{ username: string; route: string[] }>;
}) {
	const { username, route } = await params;
	const joined = route.join("/");
	const userPage = `${PATHS.user(username)}/${joined}`;
	const page = (await getPage(userPage)) as IPage;

	if (!page) {
		// omg do something
		// no page then homepage or redirect back
		return <h1>404</h1>;
	}

	return (
		<div className={styles.page}>
			<PageDisplay page={page} />
		</div>
	);
}
