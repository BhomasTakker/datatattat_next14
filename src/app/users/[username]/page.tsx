import { getPage } from "@/actions/page/page-actions";
import { IPage } from "@/types/page";
import { PageDisplay } from "@/components/page/page-display";
import styles from "../../page.module.scss";
import { PATHS } from "@/lib/routing/paths";
import { generateMetaDataFromPage } from "@/lib/metadata/generate-metadata";

type Params = Promise<{ username: string }>;
type Props = {
	params: Params;
};

export const revalidate = 600;

export const generateMetadata = async ({ params }: Props) => {
	const { username } = await params;

	const metadata = await generateMetaDataFromPage(PATHS.user(username));
	return metadata;
};

export default async function UserHome({ params }: Props) {
	const { username } = await params;

	const page = (await getPage(PATHS.user(username))) as IPage;

	if (!page) {
		// omg do something
		// no page then homepage or redirect back
		return <h1>404</h1>;
	}

	return (
		<div className={styles.page}>
			<PageDisplay page={page} />
		</div>
	);
}
