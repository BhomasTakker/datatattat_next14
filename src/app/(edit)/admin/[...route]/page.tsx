import { getServerSession } from "next-auth";
import styles from "../../../page.module.scss";
import { EditPage } from "@/components/edit/edit-page";
import { getUserById } from "@/lib/mongo/actions/user";
import { Session } from "@/types/auth/session";
import { options } from "@/app/api/auth/[...nextauth]/options";
import { PATHS } from "@/lib/routing/paths";

export const dynamic = "force-dynamic";

type PageProps = {
	params: Promise<{ route: string[] }>;
};

export default async function Page({ params }: PageProps) {
	const session = (await getServerSession(options)) as Session;
	const { user: sessionUser } = session;
	const { role } = await getUserById(sessionUser.user_id);

	// need select from multiple levels of admin?
	const adminLevel = PATHS.home();

	const { route } = await params;
	const joined = route.join("/");

	// Get/Check level(s) of access
	if (role !== "admin") {
		return <div>Unauthorized</div>;
	}

	return (
		<div className={styles.page}>
			<EditPage route={`/${joined}`} title="Admin Edit" isAdminEdit={true} />
		</div>
	);
}
