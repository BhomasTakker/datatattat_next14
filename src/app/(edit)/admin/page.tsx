import styles from "../../page.module.scss";
import { EditPage } from "@/components/edit/edit-page";
import isValidSession from "@/actions/auth/check-session";
import isSignupComplete from "@/actions/signup/signup-completed";
import { isValidUser } from "@/actions/auth/check-valid-user";
import { connectToMongoDB } from "@/lib/mongo/db";

export const dynamic = "force-dynamic";

// This is a bit of a hack page
// /admin and /admin/ are the same thing
// But we need to differentiate between the two
// So /admin is now home page admin
// /admin[route] is the admin edit page
export default async function Page() {
	await connectToMongoDB();
	await isValidSession();
	await isSignupComplete();

	const { role } = await isValidUser();

	// need select from multiple levels of admin?
	const adminLevel = "/";

	if (role !== "admin") {
		return <div>Unauthorized</div>;
	}

	return (
		<div className={styles.page}>
			<EditPage route={adminLevel} title="Admin Edit" isAdminEdit={true} />
		</div>
	);
}
