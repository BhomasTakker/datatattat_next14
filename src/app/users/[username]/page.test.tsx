import React from "react";
import { render, screen } from "@testing-library/react";
import UserHome, { generateMetadata } from "./page";
import { getPage } from "@/actions/page/page-actions";
import { PATHS } from "@/lib/routing/paths";
import { generateMetaDataFromPage } from "@/lib/metadata/generate-metadata";

// Mock dependencies
jest.mock("../../../actions/page/page-actions");
jest.mock("../../../lib/routing/paths", () => ({
	PATHS: {
		user: (username: string) => `/users/${username}`,
	},
}));
jest.mock("../../../lib/metadata/generate-metadata");
jest.mock("../../../components/page/page-display", () => ({
	PageDisplay: ({ page }: any) => (
		<div data-testid="page-display">{page?.title}</div>
	),
}));
jest.mock("../../../components/header/main-header", () => ({
	MainHeader: ({ route }: any) => (
		<div data-testid="main-header">{route.join("/")}</div>
	),
}));
jest.mock("../../page.module.scss", () => ({
	page: "mocked-page-class",
}));

describe("UserHome", () => {
	const username = "testuser";
	const params = Promise.resolve({ username });

	afterEach(() => {
		jest.clearAllMocks();
	});

	it("renders 404 if page is not found", async () => {
		(getPage as jest.Mock).mockResolvedValueOnce(null);

		// @ts-ignore
		const { findByText } = render(await UserHome({ params }));
		expect(await findByText("404")).toBeInTheDocument();
	});

	it("renders MainHeader and PageDisplay when page exists", async () => {
		const mockPage = { title: "Test Page" };
		(getPage as jest.Mock).mockResolvedValueOnce(mockPage);

		// @ts-ignore
		const { findByTestId } = render(await UserHome({ params }));
		expect(await findByTestId("main-header")).toHaveTextContent(
			"users/testuser"
		);
		expect(await findByTestId("page-display")).toHaveTextContent("Test Page");
	});

	describe("Snaps", () => {
		it("renders Page Unchanged", async () => {
			const mockPage = { title: "Test Page" };
			(getPage as jest.Mock).mockResolvedValueOnce(mockPage);
			const { container } = render(await UserHome({ params }));
			expect(container).toMatchSnapshot();
		});
	});
});

describe("generateMetadata", () => {
	it("calls generateMetaDataFromPage with correct path", async () => {
		const mockMetadata = { title: "Meta" };
		(generateMetaDataFromPage as jest.Mock).mockResolvedValueOnce(mockMetadata);

		const params = Promise.resolve({ username: "metauser" });
		const result = await generateMetadata({ params });

		expect(generateMetaDataFromPage).toHaveBeenCalledWith("/users/metauser");
		expect(result).toBe(mockMetadata);
	});
});
