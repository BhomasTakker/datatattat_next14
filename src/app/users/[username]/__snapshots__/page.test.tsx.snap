// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserHome Snaps renders Page Unchanged 1`] = `
<div>
  <div
    data-testid="main-header"
  >
    users/testuser
  </div>
  <div
    class="mocked-page-class"
  >
    <div
      data-testid="page-display"
    >
      Test Page
    </div>
  </div>
</div>
`;
